lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -222,13 +222,14 @@ def add_purchase_card(post, options)
       end
 
       def add_card_on_file(post, options)
-        return unless stored_credential = options[:stored_credential]
+        stored_credential = options[:stored_credential] || {}
 
         post[:cardOnFile] = {}
-        post[:cardOnFile][:usageIndicator] = stored_credential[:initial_transaction] ? '01' : '02'
-        post[:cardOnFile][:indicator] = options[:card_on_file_indicator] || '01'
-        post[:cardOnFile][:scheduledIndicator] = RECURRING_TYPE_TRANSACTIONS.include?(stored_credential[:reason_type]) ? '01' : '02' if stored_credential[:reason_type]
-        post[:cardOnFile][:transactionId] = stored_credential[:network_transaction_id] if stored_credential[:network_transaction_id]
+        post[:cardOnFile][:usageIndicator] = options[:usage_indicator] || (stored_credential[:initial_transaction] ? '01' : '02') if stored_credential[:reason_type] || options[:usage_indicator]
+        post[:cardOnFile][:indicator] = options[:indicator] || '01' if stored_credential[:reason_type] || options[:indicator]
+        post[:cardOnFile][:scheduledIndicator] = options[:scheduled_indicator] || (RECURRING_TYPE_TRANSACTIONS.include?(stored_credential[:reason_type]) ? '01' : '02') if stored_credential[:reason_type] || options[:scheduled_indicator]
+        post[:cardOnFile][:transactionId] = options[:transaction_id] || (stored_credential[:network_transaction_id]) if stored_credential[:network_transaction_id] || options[:transaction_id]
```

Comment:
I think you might be able to reduce your the logic here to this:
`... = options[:usage_indicator] || (stored_credential[:initial_transaction] ? '01' : '02')`.

If no `options[:usage_indicator]` is set, then it should handle the `||` part of the condition. Anything after the `if` will be ignored.

I see what you're trying to do here and I think it's good intuition, but it might make sense to validate the existence of one or the other set of fields (stored_credential vs GSFs) and return early from the method if neither exists. That said, I'm not sure how far we need to go here in preventing anything other than TypeErrors and other logic errors that might bubble up. `options[:usage_indicator]` will just return `nil` if that key isn't found, in which case the gateway might yield some error message we can pass back to the user.

Reply:
Yes my idea was to prevent it from passing nil values to the gateway in any case (even when one card on file is passed and other is not) but since all tests do pass with the changes you suggested so we should be good to apply them.

I have applied these changes with the extra test to validate the fields when GSFs and SC fields are combinedly passed 👍